File: bin/256-colors.js
#!/usr/bin/env node
// jshint esversion: 6
// ported to Node from https://gist.github.com/graven/921334
// Demonstrate using es6 generators

// print "color indexes should be drawn in bold text of the same color.\n"
function padn(num, len) { return (num+"").length >= len ? num+"" : ("00000000000000000000"+num).substr(0-len); }
function padx(num, len) { num=num.toString(16);
                          return num.length >= len ? num : ("00000000000000000000"+num).substr(0-len); }
function pads(str, len) { return str.length >= len ? str : ("                    "+str).substr(0-len); }

function* colored()  { yield(0); for (let n=0; n<=4; n++) yield(0x5f + 40 * n); }
function* gray()  { for (let n=0; n<=23; n++) yield(8 + 10 * n); }

let gen = (function*() {
	let i = 16;
	for (let r of colored()) for (let g of colored()) for (let b of colored())
		yield [i++, padx(r,2)+"/"+padx(g,2)+"/"+padx(b,2)];
	for (let a of gray())
		yield [i++, padx(a,2)+"/"+padx(a,2)+"/"+padx(a,2)];
})();

let normal = c => "\x1b[48;5;"+c+"m";
let textfg = c => "\x1b[38;5;"+c+"m";
let bold   = c => "\x1b[1;38;5;"+c+"m";
let reset  = "\x1b[0m";

function* col8ln(l) {
	for (let [i,colnm] of ["black", "red ", "green", "yellow", "blue ", "magnta", "cyan ", "white"].entries()) {
		let c = l*8 + i;
		yield bold(c)+padn(c,2)+": "+reset+normal(c)+textfg(l === 0 ? 15 : 0)+pads(colnm,6)+reset;
	}
}

function* col240ln() {
	while (true) {
		let res="  ", c, str;
		for (let i = 0; i < 6; i++) {
			try { [c, str] = gen.next().value; } catch(err) { return; }
			res += bold(c)+" "+padn(c,3)+": "+reset+normal(c)+textfg((c-16)%36 < 6 ? 15 : 0)+str+reset;
		}
		yield res;
	}
}
console.log([[...col8ln(0)].join(" "),[...col8ln(1)].join(" ")].join('\n'));
console.log([...col240ln()].join('\n'));

File: bin/256-colors.lua
#!/usr/bin/env luajit
-- ported to Lua from https://gist.github.com/graven/921334
-- Demonstrate using Lua coroutines as generators

-- print "color indexes should be drawn in bold text of the same color.\n"

local cocreate, yield, resume = coroutine.create, coroutine.yield, coroutine.resume

local function iter(co) return function()
    local r = { resume(co) }
    if r[1] and r[2] then
        return unpack(r, 2)
    end
    return nil
end end

local function colored() return cocreate(function()
   yield(0) for n=0,4 do yield(0x5f + 40 * n) end
end) end

local function gray() return cocreate(function()
    for n=0,23 do yield(8 + 10 * n) end
end) end

local gen = cocreate(function()
    local i = 16
    for r in iter(colored()) do for g in iter(colored()) do for b in iter(colored()) do
        yield(i, ("%02x/%02x/%02x"):format(r,g,b))
        i = i + 1
    end end end
    for a in iter(gray()) do
        yield(i, ("%02x/%02x/%02x"):format(a,a,a))
        i = i + 1
    end
end)

local normal = "\x1b[48;5;%sm"
local textfg = "\x1b[38;5;%sm"
local bold   = "\x1b[1;38;5;%sm"
local reset  = "\x1b[0m"

for t,typ in ipairs { "", "*" } do
    for c,colnm in ipairs {"black", "red ", "green", "yellow", "blue ", "magnta", "cyan ", "white"} do
        local i = (t-1) * 8 + c - 1
        local index = (bold.." %02d: "..reset):format(i, i)
        local color = (normal..textfg.."%6s"..reset):format(i, t == 1 and 15 or 0, colnm)
        io.stdout:write(index..color..(c==8 and '\n' or ''))
    end
end
for i, color in iter(gen) do
    local index = (bold.." %03d: "..reset):format(i, i)
    local hex   = (normal..textfg.."%s"..reset):format(i, (i-16) % 36 < 6 and 15 or 0, color)
    local indent = i % 6 == 4 and '  ' or ''
    local newline = i % 6 == 3 and '\n' or ''
    io.stdout:write(indent..index..hex..newline) 
end
